Provide a generic delete entity action.

bojanz opened the pull request:
Continued from https://www.drupal.org/node/2650152

bojanz committed:
Provide a generic delete entity action.

dawehner commented on src/Form/DeleteMultiple.php:
```diff
@@ -0,0 +1,222 @@
+   *   The tempstore factory.
+   */
+  public function __construct(AccountInterface $current_user, EntityTypeManagerInterface $entity_type_manager, PrivateTempStoreFactory $temp_store_factory) {
+    $entity_type_id = \Drupal::routeMatch()->getParameter('entity_type');
```
Do we really need to execute code in the constructor? At least we could pass in the route match in the create() method

bojanz replied:
Yeah, we definitely need to inject the route match.

Getting stuff from the entity type manager in the constructor is a pattern we use in a lot of places.

dawehner replied:
Well, IMHO its okay for plugins, but I would avoid it for services, well especially low level ones

dawehner commented on src/Form/DeleteMultiple.php:
```diff
@@ -0,0 +1,222 @@
+   * {@inheritdoc}
+   */
+  public function getConfirmText() {
+    return t('Delete');
```
$this->t()

dawehner commented on src/Form/DeleteMultiple.php:
```diff
@@ -0,0 +1,222 @@
+   * {@inheritdoc}
+   */
+  public function buildForm(array $form, FormStateInterface $form_state) {
+    $this->entityInfo = $this->tempStore->get($this->currentUser->id());
```
This is IMHO a bad var name

bojanz replied:
Yes, it's terrible. The node form used $this->nodeInfo, so I guess that just got transferred.
Ideally it wouldn't even be a class attribute, but getQuestion() is using it...

dawehner commented on src/Form/DeleteMultiple.php:
```diff
@@ -0,0 +1,222 @@
+      return new RedirectResponse($this->getCancelUrl()->setAbsolute()->toString());
+    }
+
+    /** @var \Drupal\Core\Entity\ContentEntityInterface[] $entities */
```
General question, would this class be possible for all entities, not just content entities?

bojanz replied:
No, because of the translation stuff, that's content entity specific. We'd need a different form for config entities.

dawehner replied:
Why can't things be nice :(

dawehner commented on src/Routing/DeleteMultipleRouteProvider.php:
```diff
@@ -0,0 +1,52 @@
+      $route = new Route($entity_type->getLinkTemplate('delete-multiple-form'));
+      $route->setDefault('_form', '\Drupal\entity\Form\DeleteMultiple');
+      $route->setDefault('entity_type_id', $entity_type->id());
+      $route->setRequirement('_permission', $entity_type->getAdminPermission());
```
That permissions is a bit meh, but I guess there isn't any alternative out there?

bojanz replied:
_entity_access: $entity_type . '.delete' ?

dawehner replied:
Well, the problem is that _entity_access is bound to single entities mostly, which is not the case here. Ideally we would check _entity_access: $entity_type.delete for each stored entity?

dawehner replied:
So yeah I still believe we have to check access to each individual entity. Better be safe than sorry

bojanz replied:
Actually, BulkForm does that by calling access() on the action plugin, which in our case forwards to $entity->access('delete'). Good enough?

bojanz committed:
Address dawehner's feedback.

- Rename entityInfo to selection.
- Inject the $entity_type_id to the form.
- Do less in the constructor.

bojanz committed:
Add tests.

bojanz commented:
I think this is good to go. The tests could be expanded if we add a label to the EnhancedEntity, then check for the presence of labels in DeleteMultipleFormTest.

dawehner commented:
Ship it!

dawehner commented:
Can we have a core issue for this, please :)

bojanz commented:
Done, 2670730.
